Solve race condition in password lookup

HD650 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Password lookup will face race conditions when running parallelly, this PR use atomic file creation to solve the race condition.
Fix #37116 
<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
ansible/lib/ansible/plugins/lookup/password.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (issue_37116 23c68b7ad2) last updated 2018/07/09 15:57:37 (GMT -400)
  config file = None
  configured module search path = [u'/home/zhikangzhang/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/zhikangzhang/Desktop/ansible/lib/ansible
  executable location = /home/zhikangzhang/Desktop/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

NOTE:
1. use os.open() with os.O_CREAT|os.O_EXCL to check existence
and create a lock file if not exists, it's an atomic operation
2. the fastest process will create the lock file and others will
wait until the lock file is removed
3. after the writer finished writing to the password file, all the reading
operations use built-in open so processes can read the file parallel

Commit:
Solve race condition in password lookup

NOTE:
1. use os.open() with os.O_CREAT|os.O_EXCL to check existence
and create a file if not exists, it's an atomic operation
2. os.open can't work on a not exists dir so create the dir when start
3. the fastest process will create the password file and others will
wait until there is contents appeared in the password file
4. after the creation of password file, all file operation use built-in
open so processes can read the file parallel

Commit:
Handle special case 'dev/null'

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -161,11 +162,19 @@ def _read_password_file(b_path):
+                b_content = f.read().rstrip()
+                content = to_text(b_content, errors='surrogate_or_strict')
+        else:
+            raise e
```
Just `raise`.

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -161,11 +162,19 @@ def _read_password_file(b_path):
-        with open(b_path, 'rb') as f:
-            b_content = f.read().rstrip()
-        content = to_text(b_content, errors='surrogate_or_strict')
+    fd = -1
```
rm

Commit:
Switch to lockfile

Using lockfile can remain all existing function unchanged.

mscherer commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -276,10 +278,29 @@ def run(self, terms, variables, **kwargs):
+            writer_process = False
+            if not os.path.exists(b_path):
+                try:
+                    fd = os.open("./ansible_password_lookup.lockfile", os.O_CREAT | os.O_EXCL)
```
This wouldn't work if ansible can't write in the current working directory.

HD650 replied:
I will make the lock file locates at the same directory as the password file. And the user will make sure ansible has enough privilege to create a file there.

mscherer commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -276,10 +278,29 @@ def run(self, terms, variables, **kwargs):
+                            timeout += 1
+                            if timeout > 20:
+                                raise AnsibleError("Password lookup cannot get the lock in 2 seconds, abort...\n"
+                                                   "This may caused by un-removed lockfile, you can manually remove it")
```
The error message do not tell where to remove the file, it might be a bit misleading to the person running ansible.

HD650 replied:
OK, I will add this.

mscherer commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -276,10 +278,29 @@ def run(self, terms, variables, **kwargs):
+            if not writer_process:
+                content = _read_password_file(b_path)
+
+            if writer_process or b_path == to_bytes('/dev/null'):
```
writer_process will never be true, since /dev/null is supposed to always exist.

HD650 replied:
Users may input some not exits path like `/path/to/password/password.txt`, in this case, there will be  one writer process.

mscherer replied:
Ok, yeah, I misunderstood that part.

Commit:
Place lockfile at the same directory as passwordfile

Commit:
Fix unit test error

mscherer commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -276,10 +279,32 @@ def run(self, terms, variables, **kwargs):
-
-            if content is None or b_path == to_bytes('/dev/null'):
+            writer_process = False
+            if not os.path.exists(b_path) and b_path != to_bytes('/dev/null'):
```
So, if b_path is '/dev/null', why do we have the 2nd test, and in which case would /dev/null be inexistant ?

mscherer commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -276,10 +279,32 @@ def run(self, terms, variables, **kwargs):
+                            timeout += 1
+                            if timeout > 20:
+                                raise AnsibleError("Password lookup cannot get the lock in 2 seconds, abort..."
+                                                   "This may caused by un-removed lockfile, you can manually remove it at %s and try again" % lockfile)
```
I still think the message could remind where is the file (like, on controller, vs on remote system). It should be obvious for people, yes, but not every user realize that lookup are on controller side.

HD650 replied:
Make sense, I will fix this.

mscherer commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -102,6 +104,7 @@
 
 DEFAULT_LENGTH = 20
 VALID_PARAMS = frozenset(('length', 'encrypt', 'chars'))
+LOCKFILE_NAME = to_bytes("ansible_lookup_password.lockfile")
```
Doesn't that mean that suddenly, all password lookup are locked on the same file ? If that's for 2 differents password lookup on different file, it should lock. While that's unlikely to happen, I think it would be better to avoid 1 single big lock. Could something like a hash of the file name be used and/or added ?

mscherer replied:
Also, what if a user use a password file with the name "ansible_lookup_password.lockfile"

HD650 replied:
yes make sense, I will find some way to produce a random name so workers will not block on a single lock and the name of lockfile will not be same as password file.

webknjaz replied:
Typically it's a convention to have a `.lock` extension for such files.

Commit:
Make timeout msg more clear

Commit:
Let lookups not lock on different passwordfiles

Commit:
Make lock workflow easier to read.

Commit:
Do not mix bytes and string in path

Commit:
Fix unit tests

mscherer commented:
So, I wonder if it would be better to wrap all in a try/finally, or something, but let's not perfect be the enemy of good.

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -269,30 +272,54 @@ def _write_password_file(b_path, content):
 class LookupModule(LookupBase):
     def run(self, terms, variables, **kwargs):
         ret = []
-
```
Why did you do this?

webknjaz commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -372,6 +373,14 @@ def setUp(self):
         self.fake_loader = DictDataLoader({'/path/to/somewhere': 'sdfsdf'})
         self.password_lookup = password.LookupModule(loader=self.fake_loader)
         self.os_path_exists = password.os.path.exists
+        self.os_open = password.os.open
```
Can we use `patch` instead?

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -269,30 +272,54 @@ def _write_password_file(b_path, content):
+                try:
+                    makedirs_safe(b_pathdir, mode=0o700)
+                    fd = os.open(lockfile, os.O_CREAT | os.O_EXCL)
+                    os.close(fd)
```
Why close fd?

HD650 replied:
The lock file is only a placeholder, we don't need to write any content to it so I close it immediately after creating it.

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -269,30 +272,54 @@ def _write_password_file(b_path, content):
+            timeout = 0
+            while os.path.exists(lockfile) and not writer_process:
+                timeout += 1
+                time.sleep(0.1)
```
Can/should we implement an exponential backoff algorithm with random offset?

webknjaz replied:
```python
counter += 1
max_sleep_time = 20
wait_for = max(max_sleep_time, counter ** 2 + random.random())
time.sleep(wait_for)

```

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -269,30 +272,54 @@ def _write_password_file(b_path, content):
 
-            changed = False
-            content = _read_password_file(b_path)
+            content = None
```
It might be a good idea to separate this chunk into a dedicated function for the sake of readability trying to preserve SRP.

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -269,30 +272,54 @@ def _write_password_file(b_path, content):
+                    os.close(fd)
+                    writer_process = True
+                except OSError as e:
+                    if not e.strerror == 'File exists':
```
This is a weird combo of comparison operator with a logical negation. You probably meant `if e.strerror != 'File exists':`

webknjaz commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -269,30 +272,54 @@ def _write_password_file(b_path, content):
+                    if not e.strerror == 'File exists':
+                        raise
+
+            timeout = 0
```
`aquire_password_file_lock(max_timeout)`

Commit:
Capsulate the lock code

Commit:
Support salt change without rage condition

If the password file already exist but has no salt,
it's also possible to write to the password file (add salt),
so we should be stricter, whatever the situation is,
there will be only one process run all the job and complete
all the writing first, then other processes continue to read.

Commit:
Use exponential sleep time

Commit:
Fix the '/dev/null' corner case

webknjaz commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -436,7 +449,7 @@ def test_password_already_created_encrypt(self, mock_get_paths, mock_write_file)
     def test_password_already_created_no_encrypt(self, mock_get_paths, mock_write_file):
         mock_get_paths.return_value = ['/path/one', '/path/two', '/path/three']
-        password.os.path.exists = lambda x: True
+        password.os.path.exists = lambda x: True if x == to_bytes('/path/to/somewhere') else False
```
Simplier:
```python
password.os.path.exists = lambda x: x == to_bytes('/path/to/somewhere')
```

Commit:
A unit test for lock mechanism

Commit:
Simplfy the test code

webknjaz commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -452,3 +465,25 @@ def test_only_a(self, mock_get_paths, mock_write_file):
+            with patch.object(builtins, 'open', mock_open(read_data=b'hunter42 salt=87654321\n')) as m:
+                # should timeout here
+                results = self.password_lookup.run([u'/path/to/somewhere chars=anything'], None)
+                self.assertEquals(True, False)
```
`self.fail('explanation text')`: https://stackoverflow.com/a/14768181

Commit:
Separate the testcase

webknjaz commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -466,24 +466,26 @@ def test_only_a(self, mock_get_paths, mock_write_file):
                 # should timeout here
                 results = self.password_lookup.run([u'/path/to/somewhere chars=anything'], None)
-                self.assertEquals(True, False)
+                self.fail('Lookup didnt timeout when lock already been held, testcase failed')
```
`s/didnt/didn't/`

webknjaz replied:
you don't need to say `testcase failed`, it's obvious from testrunner's indication

Commit:
Fix English wording

webknjaz commented:
@HD650 this probably deserves backporting. Please check it with RMs.

cc @nitzmahone @mattclay
